app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,322 @@
+            console.warn(heldValue.interval_name + 'was dropped without a call to `measure`.')
+        }
+    }
+})
+
+// ===================
+// === Measurement ===
+// ===================
+
+/**
+ * Profiling measurement that contains the profiling information about a measured time interval.
+ */
+class Measurement {
+    readonly startTime: number
+    readonly duration: number
+    readonly logLevel: Profiler
+    readonly name: string
+
+    constructor(startTime: number, duration: number, logLevel: Profiler, name: string) {
+        this.startTime = startTime
+        this.duration = duration
+        this.logLevel = logLevel
+        this.name = name
+    }
+
+    static fromPerformanceEntry(entry: PerformanceEntry): Measurement | undefined {
+        const startTime = entry.startTime
+        const duration = entry.duration
+
+        let decodedLabel = decodeLabel(entry.name)
+        if (decodedLabel === undefined) {
+            console.warn(`Could not decode "${entry.name}" as measurement label.`)
+            return
+        }
+        const [logLevel, name] = decodedLabel
+
+        return new Measurement(startTime, duration, logLevel, name)
+    }
+
+    endTime(): number {
+        return this.startTime + this.duration
+    }
+
+    prettyString(): string {
+        return `[${this.startTime.toFixed(2)},${this.endTime().toFixed(
+            2
+        )}] (${this.duration.toFixed(2)}) ${this.name}`
+    }
+
+    prettyPrint() {
+        console.log(this.prettyString())
+    }
+}
+
+// ==============
+// === Report ===
+// ==============
+
+/**
+ * Report of all measurements taken during the current session.
+ */
+export class Report {
+    measurements: Measurement[]
+
+    constructor() {
+        const entries = performance.getEntriesByType('measure')
+        this.measurements = entries.map(entry => Measurement.fromPerformanceEntry(entry))
+    }
+
+    singleLevelMeasurement(profiler: Profiler) {
+        return this.measurements.filter(m => m.logLevel === profiler)
+    }
+
+    /**
+     * Render the report to the developer console.
+     */
+    renderToDevConsole() {
+        console.groupCollapsed('PerformanceReport')
+
+        for (const profiler of profiler_registry) {
+            console.groupCollapsed(profiler.log_level)
+            this.singleLevelMeasurement(profiler).forEach(m => m.prettyPrint())
+            console.groupEnd()
+        }
+
+        console.groupEnd()
+    }
+
+    toString() {
+        let output = ''
+        for (const profiler of profiler_registry) {
+            this.singleLevelMeasurement(profiler).forEach(
+                m => (output = output.concat(m.prettyString(), '\n'))
+            )
+        }
+        return output
+    }
+}
+
+// ==================
+// === Public API ===
+// ==================
+
+export const section = new Profiler('Section')
+export const task = new Profiler('Task')
+export const symbol = new Profiler('Symbol')
+export const debug = new Profiler('Debug')
```

Comment:
I have a question here – I did not see this anywhere in the code, but if I set the performance logging level to "Symbol", I understand that "Debug" will be hidden, but "Symbol", "Task", and "Section" will be left, is that correct? I mean – these should be hierarchical levels and you should be able to set "what minimal level is interesting to you".

Reply:
This was an error. It should be `Section` and has been changed. Also, the hierarchical  setting is implemented.